subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/modulecache/DefaultCachedMetadata.java

```diff
@@ -25,35 +24,42 @@
 import javax.annotation.Nullable;
 import java.util.Collections;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 class DefaultCachedMetadata implements ModuleMetadataCache.CachedMetadata {
     private final ModuleSource moduleSource;
     private final long ageMillis;
     private final ModuleComponentResolveMetadata metadata;
-    private Map<Integer, ModuleComponentResolveMetadata> processedMetadataByRules;
 
-    public DefaultCachedMetadata(ModuleMetadataCacheEntry entry, ModuleComponentResolveMetadata metadata, BuildCommencedTimeProvider timeProvider) {
+    private volatile Map<Integer, ModuleComponentResolveMetadata> processedMetadataByRules;
```

Comment:
I guess you don't need `volatile` and `Concurrent` now that the whole method is synchronized, do you?

Reply:
They are needed. The `volatile` will make sure the swap of the map are visible to thread that did not try a put. And the `CHM` is required to protect readers during potential map resize.
The alternative is to also put a `synchronized` on the `get` but that seems too high a cost.